### ISSUE TITLE
Sanitizers fail a lot, known?

### ISSUE THREAD

**Issue by danlark1** [VectorCamp/vectorscan#91]
Hi, we decided to run ASAN sanitizers with vectorscan and it fails in various places

Is that a known issue? I believe we tested hyperscan and tests were generally fine, however, with some rewrites it's hard to say what changed

As an example:

```
[ RUN      ] HyperscanTestBehaviour.SerializedDogfood1
=================================================================
==3279==ERROR: AddressSanitizer: global-buffer-overflow on address 0x7f55c3b63c1f at pc 0x7f55c248feb2 bp 0x7ffcfbca82a0 sp 0x7ffcfbca8298
READ of size 16 at 0x7f55c3b63c1f thread T0
    #0 0x7f55c248feb1 in loadu [src/util/supervector/arch/x86/impl.cpp:502]:12
    #1 0x7f55c248feb1 in scanDoubleUnaligned<(unsigned short)16> [src/hwlm/noodle_engine_simd.hpp:98]:24
    #2 0x7f55c248feb1 in scanDoubleMain<(unsigned short)16> [src/hwlm/noodle_engine_simd.hpp:204]:12
    #3 0x7f55c248feb1 in scanDouble [src/hwlm/noodle_engine_simd.hpp:230]:12
    #4 0x7f55c248feb1 in scan [src/hwlm/noodle_engine.cpp:118]:16
    #5 0x7f55c248feb1 in noodExec [src/hwlm/noodle_engine.cpp:132]:12
    #6 0x7f55c2f50bdf in pureLiteralBlockExec [src/runtime.c:218]:5
    #7 0x7f55c2f50bdf in hs_scan [src/runtime.c:422]:9
    #8 0x7f55c3cf59aa in (anonymous namespace)::HyperscanTestBehaviour_SerializedDogfood1_Test::TestBody() [unit/hyperscan/behaviour.cpp:651]:11
    #9 0x7f5576ad39d4 in testing::Test::Run() [googletest/src/gtest.cc:2731]:5
    #10 0x7f5576ad64bb in testing::TestInfo::Run() [googletest/src/gtest.cc:2910]:11
    #11 0x7f5576ad882b in testing::TestSuite::Run() [googletest/src/gtest.cc:3069]:30
    #12 0x7f5576b0a124 in testing::internal::UnitTestImpl::RunAllTests() [googletest/src/gtest.cc:5942]:44
    #13 0x7f5576b08fe9 in testing::UnitTest::Run() [googletest/src/gtest.cc:5511]:10
    #14 0x55fd78fbf84f in RUN_ALL_TESTS [googletest/include/gtest/gtest.h:2326]:46
    #15 0x55fd78fbf84f in main [testing/base/internal/gunit_main.cc:83]:10

0x7f55c3b63c1f is located 33 bytes to the left of global variable '<string literal>' defined in (0x7f55c3b63c40) of size 2
  '<string literal>' is ascii string '<'
0x7f55c3b63c1f is located 7 bytes to the right of global variable '<string literal>' defined in '[unit/hyperscan/behaviour.cpp:646]:24' (0x7f55c3b63c00) of size 24
  '<string literal>' is ascii string 'delicious puppy treats!'
SUMMARY: AddressSanitizer: global-buffer-overflow [src/util/supervector/arch/x86/impl.cpp:502]:12 in loadu
Shadow bytes around the buggy address:
  0x0feb38764730: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
  0x0feb38764740: 07 f9 f9 f9 f9 f9 f9 f9 00 00 00 00 00 00 00 00
  0x0feb38764750: 01 f9 f9 f9 f9 f9 f9 f9 00 06 f9 f9 00 06 f9 f9
  0x0feb38764760: 00 00 01 f9 f9 f9 f9 f9 07 f9 f9 f9 00 07 f9 f9
  0x0feb38764770: 00 f9 f9 f9 00 01 f9 f9 00 00 00 07 f9 f9 f9 f9
=>0x0feb38764780: 00 00 00[f9]f9 f9 f9 f9 02 f9 f9 f9 00 00 00 00
  0x0feb38764790: 00 00 00 00 00 00 00 00 07 f9 f9 f9 f9 f9 f9 f9
  0x0feb387647a0: 00 00 00 00 00 00 00 00 01 f9 f9 f9 f9 f9 f9 f9
  0x0feb387647b0: 00 00 04 f9 f9 f9 f9 f9 00 00 04 f9 f9 f9 f9 f9
  0x0feb387647c0: 04 f9 f9 f9 00 05 f9 f9 00 03 f9 f9 07 f9 f9 f9
  0x0feb387647d0: 00 04 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==3279==ABORTING
```

**Comment by danlark1** [comment]
A couple more

```
[ RUN      ] regression.UE_2798
==3279==ERROR: AddressSanitizer: global-buffer-overflow on address 0x7f1f4cf65fd1 at pc 0x7f1f4bbdea90 bp 0x7ffe7122f2f0 sp 0x7ffe7122f2e8
READ of size 16 at 0x7f1f4cf65fd1 thread T0
    #0 0x7f1f4bbdea8f in loadu_maskz src/util/supervector/arch/x86/impl.cpp:518:21
    #1 0x7f1f4bbdea8f in nvermicelliExecReal<(unsigned short)16> src/nfa/vermicelli_simd.cpp:173:31
    #2 0x7f1f4bbdea8f in nvermicelliExec src/nfa/vermicelli_simd.cpp:499:12
    #3 0x7f1f4bb66140 in limitByReach src/nfa/mpv.c:262:16
    #4 0x7f1f4bb66140 in handleTopN src/nfa/mpv.c:391:16
    #5 0x7f1f4bb5ce32 in nfaExecMpv_Q_i src/nfa/mpv.c:1002:13
    #6 0x7f1f4bb5ce32 in nfaExecMpv_QueueExecRaw src/nfa/mpv.c:1069:15
    #7 0x7f1f4bf796ab in roseCatchUpMPV_i src/rose/catchup.c:403:26
    #8 0x7f1f4bff1bfc in roseCatchUpTo src/rose/catchup.h:158:14
    #9 0x7f1f4bff1bfc in roseRunProgram src/rose/program_runtime.c:2500:21
    #10 0x7f1f4bfbcec4 in roseProcessMatchInline src/rose/match.c:244:16
    #11 0x7f1f4bfbcec4 in roseCallback_i src/rose/match.c:512:10
    #12 0x7f1f4bfbcec4 in roseFloatingCallback src/rose/match.c:533:12
    #13 0x7f1f4b849c4c in confWithBit src/fdr/fdr_confirm_runtime.h:96:20
    #14 0x7f1f4b849c4c in do_confWithBit_teddy src/fdr/teddy_runtime_common.h:438:9
    #15 0x7f1f4b849c4c in fdr_exec_teddy_msks3 src/fdr/teddy.c:1097:5
    #16 0x7f1f4b7e47f0 in fdrExecStreaming src/fdr/fdr.c:851:15
    #17 0x7f1f4b8886ca in hwlmExecStreaming src/hwlm/hwlm.c:246:12
    #18 0x7f1f4c33cf1b in roseStreamExec src/rose/stream.c:663:9
    #19 0x7f1f4c354e32 in rawStreamExec src/runtime.c:792:5
    #20 0x7f1f4c354e32 in hs_scan_stream_internal src/runtime.c:945:9
    #21 0x7f1f4c35450f in hs_scan_stream src/runtime.c:992:21
    #22 0x7f1f4d137b36 in (anonymous namespace)::regression_UE_2798_Test::TestBody() unit/hyperscan/behaviour.cpp:1490:11
    #23 0x7f1effed39d4 in testing::Test::Run() third_party/googletest/googletest/src/gtest.cc:2731:5
    #24 0x7f1effed64bb in testing::TestInfo::Run() third_party/googletest/googletest/src/gtest.cc:2910:11
    #25 0x7f1effed882b in testing::TestSuite::Run() third_party/googletest/googletest/src/gtest.cc:3069:30
    #26 0x7f1efff0a124 in testing::internal::UnitTestImpl::RunAllTests() third_party/googletest/googletest/src/gtest.cc:5942:44
    #27 0x7f1efff08fe9 in testing::UnitTest::Run() third_party/googletest/googletest/src/gtest.cc:5511:10
    #28 0x55bf5a7bf84f in RUN_ALL_TESTS third_party/googletest/googletest/include/gtest/gtest.h:2326:46
    #29 0x55bf5a7bf84f in main testing/base/internal/gunit_main.cc:83:10

0x7f1f4cf65fd1 is located 47 bytes to the left of global variable '<string literal>' defined in 'unit/hyperscan/behaviour.cpp:1498:5' (0x7f1f4cf66000) of size 79
  '<string literal>' is ascii string 'find(c.matches.begin(), c.matches.end(), MatchRecord(7, 1)) != c.matches.end()'
0x7f1f4cf65fd1 is located 15 bytes to the left of global variable '<string literal>' defined in 'unit/hyperscan/behaviour.cpp:1497:5' (0x7f1f4cf65fe0) of size 3
  '<string literal>' is ascii string '4U'
0x7f1f4cf65fd1 is located 8 bytes to the right of global variable 'dataA' defined in 'unit/hyperscan/behaviour.cpp:1484:16' (0x7f1f4cf65fc0) of size 9
  'dataA' is ascii string 'ab_baab
'
SUMMARY: AddressSanitizer: global-buffer-overflow src/util/supervector/arch/x86/impl.cpp:518:21 in loadu_maskz
Shadow bytes around the buggy address:
  0x0fe4699e4ba0: 00 00 00 00 00 00 00 00 00 00 f9 f9 f9 f9 f9 f9
  0x0fe4699e4bb0: 00 00 00 00 00 02 f9 f9 f9 f9 f9 f9 07 f9 f9 f9
  0x0fe4699e4bc0: 03 f9 f9 f9 00 00 00 00 00 00 00 00 00 07 f9 f9
  0x0fe4699e4bd0: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 f9 f9
  0x0fe4699e4be0: f9 f9 f9 f9 00 00 00 00 00 02 f9 f9 f9 f9 f9 f9
=>0x0fe4699e4bf0: 00 06 f9 f9 00 03 f9 f9 00 01[f9]f9 03 f9 f9 f9
  0x0fe4699e4c00: 00 00 00 00 00 00 00 00 00 07 f9 f9 f9 f9 f9 f9
  0x0fe4699e4c10: 00 00 00 00 00 00 00 00 00 07 f9 f9 f9 f9 f9 f9
  0x0fe4699e4c20: 00 00 00 00 00 00 00 00 00 07 f9 f9 f9 f9 f9 f9
  0x0fe4699e4c30: 00 00 00 00 00 00 00 00 00 07 f9 f9 f9 f9 f9 f9
  0x0fe4699e4c40: 00 00 00 00 00 00 00 00 00 07 f9 f9 f9 f9 f9 f9
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==3279==ABORTING
```

**Comment by markos** [comment]
Thank you, these are really helpful, I was planning to add ASAN myself but it seems there is important reason to do that sooner rather than later. I am not sure these are false positives, so will need to investigate further.

**Comment by markos** [comment]
Merged #93 so closing this.
